src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -99,28 +156,147 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
+                await UpdateJobAsync(cancellationToken);
+
+                await CheckJobCompletionStatus(cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                query.Error = ex.Message;
+
+                query.FailureCount++;
+
+                _logger.LogError(ex, $"Encountered an unhandled exception. The qquery failure count increased to {_reindexJobRecord.FailureCount}.");
+
+                if (query.FailureCount >= _reindexJobConfiguration.ConsecutiveFailuresThreshold)
+                {
+                    query.Status = OperationStatus.Failed;
+                }
+                else
+                {
+                    query.Status = OperationStatus.Queued;
+                }
+
+                await UpdateJobAsync(cancellationToken);
+            }
+        }
+
+        private async Task CheckJobCompletionStatus(CancellationToken cancellationToken)
+        {
+            // If any query still in progress then we are not done
+            if (_reindexJobRecord.QueryList.Where(q =>
+                q.Status == OperationStatus.Queued ||
+                q.Status == OperationStatus.Running).Any())
+            {
+                return;
+            }
+            else
+            {
+                // all queries marked as complete, reindex job is done, check success or failure
+                if (_reindexJobRecord.QueryList.All(q => q.Status == OperationStatus.Completed))
+                {
+                    await CompleteJobAsync(OperationStatus.Completed, cancellationToken);
+                    _logger.LogTrace("Successfully completed the job.");
+                }
+                else
+                {
+                    await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+                    _logger.LogTrace("Reindex job did not complete successfully.");
+                }
+            }
+        }
+
+        private async Task<SearchResult> ExecuteReindexQueryAsync(ReindexJobQueryStatus queryStatus, bool countOnly, CancellationToken cancellationToken)
+        {
+            var queryParametersList = new List<Tuple<string, string>>()
+            {
+                Tuple.Create(KnownQueryParameterNames.Count, _reindexJobConfiguration.MaximumNumberOfResourcesPerQuery.ToString(CultureInfo.InvariantCulture)),
+                Tuple.Create(KnownQueryParameterNames.Type, _reindexJobRecord.ResourceList),
+            };
+
+            if (queryStatus.ContinuationToken != null)
+            {
+                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.ContinuationToken, queryStatus.ContinuationToken));
+            }
+
+            using (IScoped<ISearchService> searchService = _searchServiceFactory())
+            {
+                try
+                {
+                    return await searchService.Value.SearchForReindexAsync(queryParametersList, _reindexJobRecord.Hash, countOnly, cancellationToken);
```

Comment:
Just to confirm, we are searching for resources that have the "existing/old" search parameter hash?